packages/auth/src/providers/cognito/utils/signInHelpers.ts

```
@@ -154,49 +154,110 @@ export async function handleMFASetupChallenge({
 	config,
 }: HandleAuthChallengeRequest): Promise<RespondToAuthChallengeCommandOutput> {
 	const { userPoolId, userPoolClientId, userPoolEndpoint } = config;
-	const challengeResponses = {
+
+	if (challengeResponse === 'EMAIL') {
+		return {
+			ChallengeName: 'MFA_SETUP',
+			Session: session,
+			ChallengeParameters: {
+				MFAS_CAN_SETUP: '["EMAIL_OTP"]',
+			},
+			$metadata: {},
+		};
+	}
+
+	if (challengeResponse === 'TOTP') {
+		return {
+			ChallengeName: 'MFA_SETUP',
+			Session: session,
+			ChallengeParameters: {
+				MFAS_CAN_SETUP: '["SOFTWARE_TOKEN_MFA"]',
+			},
+			$metadata: {},
+		};
+	}
+
+	const challengeResponses: Record<string, string> = {
 		USERNAME: username,
 	};
-	const verifySoftwareToken = createVerifySoftwareTokenClient({
-		endpointResolver: createCognitoUserPoolEndpointResolver({
-			endpointOverride: userPoolEndpoint,
-		}),
-	});
-	const { Session } = await verifySoftwareToken(
-		{
-			region: getRegionFromUserPoolId(userPoolId),
-			userAgentValue: getAuthUserAgentValue(AuthAction.ConfirmSignIn),
-		},
-		{
-			UserCode: challengeResponse,
+
+	const isTOTPCode = /^\d+$/.test(challengeResponse.trim());
```

Should the trimmed string be used from this point onward?

---

Yeah that's not a bad idea. My thought was to "stay out of the way" and let the service throw the error if their input is malformed. My intent is just validate the input enough that we follow the correct code path.

---

@jjarvisp  i think that makes sense probably better to do it that way